flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java

```diff
@@ -0,0 +1,224 @@
+
+package org.apache.iceberg.flink.source;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.flink.api.common.io.DefaultInputSplitAssigner;
+import org.apache.flink.api.common.io.InputFormat;
+import org.apache.flink.api.common.io.RichInputFormat;
+import org.apache.flink.api.common.io.statistics.BaseStatistics;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.GlobalConfiguration;
+import org.apache.flink.core.io.InputSplitAssigner;
+import org.apache.flink.runtime.util.HadoopUtils;
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Flink Iceberg {@link InputFormat}.
+ * - Calls that occur on the Job manager side: {@link #createInputSplits} and {@link #getInputSplitAssigner}.
+ * - Calls that occur on the Task side: {@link #open}, {@link #reachedEnd}, {@link #nextRecord} and {@link #close}.
+ */
+public class FlinkInputFormat extends RichInputFormat<RowData, FlinkInputSplit> {
+
+  private final TableLoader tableLoader;
+  private final Schema tableSchema;
+  private final List<String> projectedFields;
+  private final ScanOptions options;
+  private final List<Expression> filterExpressions;
+  private final FileIO io;
+  private final EncryptionManager encryption;
+
+  private transient RowDataReader reader;
+  private transient boolean fetched;
+  private transient boolean hasNext;
+
+  private FlinkInputFormat(
+      TableLoader tableLoader, Schema tableSchema, FileIO io, EncryptionManager encryption,
+      List<String> projectedFields, List<Expression> filterExpressions, ScanOptions options) {
+    this.tableLoader = tableLoader;
+    this.tableSchema = tableSchema;
+    this.projectedFields = projectedFields;
+    this.options = options;
+    this.filterExpressions = filterExpressions;
+    this.io = io;
+    this.encryption = encryption;
+  }
+
+  @VisibleForTesting
+  Schema getTableSchema() {
+    return tableSchema;
+  }
+
+  @VisibleForTesting
+  List<String> getProjectedFields() {
+    return projectedFields;
+  }
+
+  @Override
+  public BaseStatistics getStatistics(BaseStatistics cachedStatistics) {
+    // Legacy method, not be used.
+    return null;
+  }
+
+  @Override
+  public FlinkInputSplit[] createInputSplits(int minNumSplits) throws IOException {
+    // Invoked by Job manager, so it is OK to load table from catalog.
+    tableLoader.open(HadoopUtils.getHadoopConfiguration(GlobalConfiguration.loadConfiguration()));
```

Comment:
Q:   Are we binding the iceberg table's hadoop configuration with flink job manager's hadoop configuration ?  Is it possible to access an iceberg table in the hadoop cluster which is different with flink's hadoop cluster ?  Seems a more reasonable way is:  Passing a customized `SerializeableConfiguration` from client, then job manager could access any hadoop clusters.

Reply:
I'm OK to pass a configuration to `FlinkInputFormat`, although there will be some serialization cost.